generate_crossword/step_10/lib/model.dart

```diff
@@ -0,0 +1,724 @@
+    required int height,
+    Iterable<CrosswordWord>? words,
+  }) {
+    return Crossword((b) {
+      b
+        ..width = width
+        ..height = height;
+      if (words != null) {
+        b.words.addAll(words);
+      }
+    });
+  }
+
+  /// Constructor for [Crossword].
+  /// Use [Crossword.crossword] instead.
+  factory Crossword([Function(CrosswordBuilder)? updates]) = _$Crossword;
+  Crossword._();
+}
+
+/// A work queue for a worker to process. The work queue contains a crossword
+/// and a list of locations to try, along with candidate words to add to the
+/// crossword.
+abstract class WorkQueue implements Built<WorkQueue, WorkQueueBuilder> {
+  static Serializer<WorkQueue> get serializer => _$workQueueSerializer;
+
+  /// The crossword the worker is working on
+  Crossword get crossword;
+
+  /// The outstanding queue of locations to try
+  BuiltMap<Location, Direction> get locationsToTry;
+
+  /// Known bad locations
+  BuiltSet<Location> get badLocations;
+
+  /// The list of unused candidate words that can be added to this crossword.
+  BuiltSet<String> get candidateWords;
+
+  /// Returns true if the work queue is complete.
+  bool get isCompleted => locationsToTry.isEmpty || candidateWords.isEmpty;
+
+  /// Create a work queue from a crossword.
+  static WorkQueue from({
+    required Crossword crossword,
+    required Iterable<String> candidateWords,
+    required Location startLocation,
+  }) =>
+      WorkQueue((b) {
+        if (crossword.words.isEmpty) {
+          // Strip candidate words too long to fit in the crossword
+          b.candidateWords.addAll(candidateWords
+              .where((word) => word.characters.length <= crossword.width));
+
+          b.crossword.replace(crossword);
+
+          b.locationsToTry.addAll({startLocation: Direction.across});
+        } else {
+          // Assuming words have already been stripped to length
+          b.candidateWords.addAll(
+            candidateWords.toBuiltSet().rebuild(
+                (b) => b.removeAll(crossword.words.map((word) => word.word))),
+          );
+          b.crossword.replace(crossword);
+          crossword.characters
+              .rebuild((b) => b.removeWhere((location, character) {
+                    if (character.acrossWord != null &&
+                        character.downWord != null) {
+                      return true;
+                    }
+                    final left = crossword.characters[location.left];
+                    if (left != null && left.downWord != null) return true;
+                    final right = crossword.characters[location.right];
+                    if (right != null && right.downWord != null) return true;
+                    final up = crossword.characters[location.up];
+                    if (up != null && up.acrossWord != null) return true;
+                    final down = crossword.characters[location.down];
+                    if (down != null && down.acrossWord != null) return true;
```

Comment:
Not necessarily saying to make this change, just including it as a potential alternative if you're trying to show off some new syntax:

```suggestion

                    if (crossword.characters[location.left] case final left?
                        when left.downWord != null) return true;
                    if (crossword.characters[location.right] case final right?
                        when right.downWord != null) return true;
                    if (crossword.characters[location.up] case final up?
                        when up.acrossWord != null) return true;
                    if (crossword.characters[location.down] case final down?
                        when down.acrossWord != null) return true;
```

Reply:
I'm not sure how I feel about that syntax =)

Reply:
Perhaps it needs more time to become standard and isn't the best for the target audience.

Probably better to avoid it for now :)